README
bogom - simple sendmail milter to interface bogofilter
Copyright (C) 2004-2007 Juan J. Martinez <jjm@usebox.net>


Required
--------

This milter requires libmiter API for sendmail 8.13.x or later.
Installing bogofilter is a good idea also.

Bogom has been developed and tested with the following:

	sendmail 8.13.0 and 8.14.1
	bogofilter 0.92.8 (with BerkeleyDB 4.2.52)


Build and Install
-----------------

This milter has been developed in OpenBSD, so you may need to tweak the
Makefile in order to compile it in other systems.

Check README.* files for specific notes.

On most Linux and BSD based systems, simply try:

$ make
$ su
# make install

Bogom has been packaged for FreeBSD (mail/milter-bogom) and non officially for
some Linux distributions.

IMPORTANT: since 1.8.1+ the bogom binary is installed into $(PREFIX)/libexec,
           so be sure to remove $(PREFIX)/sbin/bogom from any previous
           installation.


Unprivileged user
-----------------

When root starts bogom, the program drops its privileges to another user. It's
a good idea you create a new user to run the milter. By default bogofilter user
is expected.

# mkdir /var/spool/bogofilter

Configure bogofilter with:

bogofilter_dir=/var/spool/bogofilter

... and create your intial words database. 

After that, setup the environment with:

# chown -R bogofilter:bogofilter /var/spool/bogofilter
# chmod 700 /var/spool/bogofilter
# chmod 600 /var/spool/bogofilter/wordlist.db


Custom installation step by step
--------------------------------

bogom provides some defaults, but don't have to stick into them. 

Here follows a 'step by step' quickstart to configure bogom and bogofilter.

+ Install bogofilter
+ Install bogom
+ Create the '_bogom' user and group
+ Create the directories and apply permissions:

# mkdir -p /var/run/bogom && chown _bogom:_bogom /var/run/bogom
# mkdir -p /var/db/bogofilter && chown _bogom:_bogom /var/db/bogofilter

+ Copy the example bogofilter.cf file into /etc and edit it:

bogofilter_dir=/var/db/bogofilter

+ Setup bogofilter database. About 500 messages of spam/ham will be nice:

# su _bogom -c "bogofilter -s < spam.mbox"
# su _bogom -c "bogofilter -n < ham.mbox"

(use su with -m option if _bogom user doesn't have shell account)

+ Install the example conf file at /etc/bogom.conf and edit it. At least
you should set up:

user "_bogom"
pidfile "/var/run/bogom/bogom.pid"
connection "unix:/var/run/bogom/milter.sock"

+ Add to your sendmail mc file:

INPUT_MAIL_FILTER(`bogom',\
`S=unix:/var/run/bogom/milter.sock, T=S:30s;R:1m')

(rebuild and install the cf file)

+ Add bogom to your rc or init.d scripts
+ Restart sendmail
+ Exec bogom or reboot the system


* EOF *



CHANGELOG
2010-01-23  Juan J. Martínez <reidrac@usebox.net>

	* bogom.8, conf.c, conf.h, milter.c: removed rcsid

2010-01-22  Juan J. Martínez <reidrac@usebox.net>

	* CHANGELOG: generated from git log

2010-01-22  Juan J. Martínez <reidrac@usebox.net>

	* .gitignore, CHANGELOG, COPYING, Makefile, README, README.solaris,
	TODO, bogom.8, bogom.conf-example, conf.c, conf.h, milter.c: RCS
	imported tree

2007-02-08  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: copyright year

2006-10-22  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: smpamicity support for the X-Bogosity header

2005-09-23  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: Added quarantine_mdir feature (-q option)

2005-09-12  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: Added forward_spam feature (-f option)

2005-04-04  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: new configuration option 'subject_tag'

2005-02-15  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: option changes: '-p pipe' becomes '-s conn' and '-p
	pidfile' has been added (plus pidfile conf token)

2005-02-04  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: seems openbsd doesn't support !prog1,prog2 on syslog.conf

2005-01-31  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: some fixes

2005-01-27  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: -d option added and syslog.conf tips little rewrite of
	some parts

2005-01-27  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: added body_limit option

2005-01-25  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: re_envrcpt added

2005-01-22  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: -w option removed, re_connection added, year on copyright
	line

2005-01-08  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: fix: configuration takes precedence over command line
	fix: backslash escaped quotes now works (both \" and \') Quoted
	string now can use single quotes also

2005-01-04  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: -w option deprecated (now configuration file must be
	used).

2005-01-02  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: Configurable text for the SMTP reject reply.

2005-01-02  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: Added -c conf_file option.  Configuration file
	description.

2004-12-29  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: fix: D1 macros

2004-12-29  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: -w re_whitelist option added

2004-12-28  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: Added sendmail setup information.

2004-12-28  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: Better description of logging.

2004-12-27  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: -x exclude_string added

2004-12-27  Juan J. Martinez <jjm@usebox.net>

	* bogom.8: Initial revision

2007-02-08  Juan J. Martinez <jjm@usebox.net>

	* conf.h: copyright year

2005-01-23  Juan J. Martinez <jjm@usebox.net>

	* conf.h: year on copyright line

2005-01-10  Juan J. Martinez <jjm@usebox.net>

	* conf.h: Macro fixes and little code clean.

2005-01-02  Juan J. Martinez <jjm@usebox.net>

	* conf.h: Initial revision

2006-10-22  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: smpamicity support for the X-Bogosity header

2005-09-23  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: Added quarantine_mdir feature (-q option)

2005-09-12  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: Added forward_spam feature (-f option)

2005-04-04  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: new configuration option 'subject_tag'

2005-02-15  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: pidfile added

2005-01-27  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: added body_limit option

2005-01-25  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: re_envrcpt example added

2005-01-22  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: re_connection added

2005-01-03  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: fix: typo on exclude_string examples.

2005-01-02  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: Configurable text for the SMTP reject reply.

2005-01-02  Juan J. Martinez <jjm@usebox.net>

	* bogom.conf-example: Initial revision

2008-06-25  Juan J. Martinez <jjm@usebox.net>

	* milter.c: FIX: hostaddr may be NULL if not supported by current
	version of milter, noticed using milter with postfix - Thanks to
	Victor Balada Diaz

2008-01-18  Juan J. Martinez <jjm@usebox.net>

	* milter.c: fix: added an include to avoid a warning while compiling
	under Linux

2007-11-24  Juan J. Martinez <jjm@usebox.net>

	* milter.c: portability: extra flag in some stdio functions to avoid
	file descriptor limit issue on Solaris 10, thanks to Sergey
	Shapovalov

2007-02-28  Juan J. Martinez <jjm@usebox.net>

	* milter.c: Fix: space missing in a log entry.

2007-02-08  Juan J. Martinez <jjm@usebox.net>

	* milter.c: copyright year

2007-02-04  Juan J. Martinez <jjm@usebox.net>

	* milter.c: Error handling while reading bogofilter spamicity
	response.

2006-10-22  Juan J. Martinez <jjm@usebox.net>

	* milter.c: smpamicity support for the X-Bogosity header

2006-10-08  Juan J. Martinez <jjm@usebox.net>

	* milter.c: Avoid multiple X-Bogosity headers.

2006-01-05  Juan J. Martinez <jjm@usebox.net>

	* milter.c: fix: xxfi_header is called *zero* or more times (check
	http://www.usebox.net/jjm/bogom/errata/bogom-errata-2006-1.txt)

2005-11-19  Juan J. Martinez <jjm@usebox.net>

	* milter.c: fix: include changes to support MAXHOSTNAMELEN constant

2005-11-06  Juan J. Martinez <jjm@usebox.net>

	* milter.c: sprintf to snprintf

2005-09-23  Juan J. Martinez <jjm@usebox.net>

	* milter.c: Added quarantine_mdir feature (-q option)

2005-09-12  Juan J. Martinez <jjm@usebox.net>

	* milter.c: Better logging for forward_spam and subj_tag.

2005-09-12  Juan J. Martinez <jjm@usebox.net>

	* milter.c: Added forward_spam feature (-f option)

2005-04-14  Juan J. Martinez <jjm@usebox.net>

	* milter.c: fix: wrong size on calloc, thanks to Jason M. for his

conf.h
/*
* conf.h, configuration reader and parser include
* Copyright (C) 2004-2007 Juan J. Martinez <jjm*at*usebox*dot*net>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License Version 2 as
* published by the Free Software Foundation.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
*/

#ifndef __BOGOM_CONF__
#define __BOGOM_CONF__

#define REQ_NONE	0
#define REQ_BOOL	1
#define REQ_STRING	2
#define REQ_QSTRING	3
#define REQ_LSTQSTRING	4

struct string_list
{
	char *s;
	struct string_list *n;
};

struct conftoken
{
	char *word;
	int required;
	char *str;
	int bool;
	struct string_list *sl;
};

int read_conf(const char *filename, struct conftoken *);

#endif



conf.c
/*
* conf.c, configuration reader and parser
* Copyright (C) 2004-2007 Juan J. Martinez <jjm*at*usebox*dot*net>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License Version 2 as
* published by the Free Software Foundation.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*
*/

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<errno.h>
#include<ctype.h>

#include "libmilter/mfapi.h"
#include "conf.h"

#define new_string_list(x) do {\
		x=(struct string_list *) \
			malloc(sizeof(struct string_list));\
		x->n=NULL;\
	} while(0)

static char * pstrncpy(char *, const char *, size_t);
static int parse_string(char *p);
static int parse_qstring(char *p);
static int parse_bool(char *p);
static char * parse_conf(struct conftoken *conf, char *p);

/*
* strncpy alike function that parses scaped quotes
*/
static char *
pstrncpy(char *d, const char *s, size_t l)
{
	size_t i, j;

	for(i=0, j=0; i<l && s[i]; i++, j++)
	{
		if(s[i]=='\\' && i+1<l)
			if(s[i+1]=='\'' || s[i+1]=='\"')
				i++;
		d[j]=s[i];
	}

	d[j]=0;

	return d;
}

/*
* parses a string and returns its length
*/
static int
parse_string(char *p)
{
	char *t;

	for(t=p; *t && !isspace(*t); t++);

	return (int)(t-p);
}

/*
* parses a quoted string and returns its length
* on error:
*	-1	close quote expected
*	0	empty quotes
*/
static int
parse_qstring(char *p)
{
	char *t;
	char q=*p;

	for(t=++p; *t; t++)
	{
		if(*t==q)
			break;

		if(*t=='\\' && t[1]==q)
			t++;
	}

	if(*t!=q)
		return -1;

	if(t==p)
		return 0;

	return (int)(t-p);
}

/*
* parses a bool and returns 0/1
* all values ne 0 are true
*/
static int
parse_bool(char *p)
{
	if(!p[0] || !isspace(p[1]))
		return -1;

	if(*p=='0')
		return 0;
	else
		return 1;
}

/*
* parses one line each call
*/
static char *
parse_conf(struct conftoken *conf, char *p)
{
	int i, len;
	struct string_list *t;

	if(!p)
		return NULL;

	if(!p[0] || p[0]=='\n' || p[0]=='#')
		return NULL;

	while(isspace(*p))
		p++;

	len=parse_string(p);
	if(!len)
		return NULL;

	for(i=0; conf[i].word; i++)
		if(!strncmp(conf[i].word, p, strlen(conf[i].word))) 
		{
			p+=len;
			while(isspace(*p))
				p++;

			switch(conf[i].required)
			{
				case REQ_NONE:
					/* nothing required */
					break;

				case REQ_BOOL:
					len=parse_bool(p);
					if(len<0)
					{
						fprintf(stderr, 
							"bool expected\n");
						return p;
					}

					conf[i].bool=len;
					p++;
					break;

				case REQ_STRING:
					len=parse_string(p);
					if(!len)
					{
						fprintf(stderr, 
							"string expected\n");
						return p;
					}

					if(conf[i].str)
						free(conf[i].str);
					conf[i].str=(char *)malloc(len+1);
					if(!conf[i].str)
					{
						fprintf(stderr, "malloc\n");
						return p;
					}
					strncpy(conf[i].str, p, len);
					p+=len;
					break;

				case REQ_QSTRING:
					if(*p!='\"' && *p!='\'')
					{
						fprintf(stderr,
							"quoted string"
							" expected\n");
						return p;
					}

					len=parse_qstring(p);
					p++;

					if(len==-1)
					{
						fprintf(stderr, 
							"end quote expected\n");
						return p;
					}

					if(!len)
					{
						fprintf(stderr, 
							"empty quotes\n");
						return p;
					}

					if(conf[i].str)
						free(conf[i].str);
					conf[i].str=(char *)malloc(len+1);
					if(!conf[i].str)
					{
						fprintf(stderr, "malloc\n");
						return p;
					}
					pstrncpy(conf[i].str, p, len);
					p+=len+1;
					break;

				case REQ_LSTQSTRING:
					if(*p!='\"' && *p!='\'')
					{
						fprintf(stderr,
							"quoted string"
							" expected\n");
						return p;
					}

					len=parse_qstring(p);
					p++;

					if(len==-1)
					{
						fprintf(stderr, 
							"end quote expected\n");
						return p;
					}

					if(!len)
					{
						fprintf(stderr, 
							"empty quotes\n");
						return p;
					}

					if(!conf[i].sl)
					{
						new_string_list(conf[i].sl);
						if(!conf[i].sl)
						{
							fprintf(stderr,
								"malloc");
							return p;
						}
					}
					else
					{
						new_string_list(t);
						if(!t)
						{
							fprintf(stderr,
								"malloc");
							return p;
						}
						t->n=conf[i].sl;
						conf[i].sl=t;
					}

					conf[i].sl->s=(char *)malloc(len+1);
					if(!conf[i].sl->s)
					{
						fprintf(stderr, "malloc\n");
						return p;
					}
					pstrncpy(conf[i].sl->s, p, len);
					p+=len+1;
					break;
			}
			break;
		}

	if(conf[i].word)
	{
		while(isspace(*p))
			p++;

		if(!p[0])
			return NULL;
	
		fprintf(stderr, "parse error\n");
		return p;
	}

	fprintf(stderr, "unknown token\n");

	p[len]=0;
	return p;
}

/*
* reads and parses the configuration file
*/
int
read_conf(const char *filename, struct conftoken *conf)
{
	FILE *fd;
	char buffer[1024];
	char *ret;
	int line, i;

	fd=fopen(filename, "r");
	if(!fd)
		return 0;

	line=1;
	while(!feof(fd))
	{
		i=0;
		*buffer=0;
		do
		{
			if(i>1023)
			{
				fclose(fd);
				fprintf(stderr, "conf line %i is too long\n",
					 line);
				return 1;
			}

			fscanf(fd, "%c", &buffer[i]);

		} while(buffer[i++]!='\n' && !feof(fd));

		buffer[i]=0;

		ret=parse_conf(conf, buffer);
		if(ret)
		{
			fclose(fd);
			fprintf(stderr, "conf error at line %i, near: %s\n",
				line, ret);
			return 1;
		}

		line++;
	}

	fclose(fd);

	return 0;
}

/* EOF */
